[Inductor] Support custom op in JIT with cpp wrapper

desertfire opened the pull request:
Summary:  To call custom ops in an ABI-compatible way requires doing boxed call with varargs across C shim. In the JIT mode, we can get around it by calling into Python.  https://gist.github.com/desertfire/be2a65b0a9b47780bb716b53ac2cd2b3 is an example of generated code.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #122554



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @amjames @chauhang

Differential Revision: [D55326556](https://our.internmc.facebook.com/intern/diff/D55326556)

desertfire committed:
[Inductor] Support custom op in cpp wrapper codegen

[ghstack-poisoned]

desertfire committed:
Update on "[Inductor] Support custom op in cpp wrapper codegen"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames chauhang

[ghstack-poisoned]

desertfire committed:
Update on "[Inductor] Support custom op in cpp wrapper codegen"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames chauhang

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/cpp_wrapper_cpu.py:
```diff
@@ -999,6 +1000,14 @@ def generate_before_suffix(self, result):
+                    with result.indent():
+                        result.splice(
+                            """
+                            Py_DECREF(codecache_module);
```
Should we add a small RAII wrapper for all the decrefs?

desertfire committed:
Update on "[Inductor] Support custom op in JIT with cpp wrapper"

Summary:  To call custom ops in an ABI-compatible way requires doing boxed call with varargs across C shim. In the JIT mode, we can get around it by calling into Python.  https://gist.github.com/desertfire/61823fd45f97d3e8d44a22f0a364b210 is an example of generated code.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler amjames chauhang

[ghstack-poisoned]

desertfire commented:
Update with a `RAIIPyObject` class. Updated the gist link in the comment to https://gist.github.com/desertfire/be2a65b0a9b47780bb716b53ac2cd2b3

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D55326556).

chenyang78 reviewed (approved):
LGTM. Thanks!

desertfire commented:
@pytorchbot merge -f "landed internally, somehow the auto merge here was not triggered"

jeffdaily commented:
This did break ROCm inductor CI.  I opened a DISABLED issue to skip.  https://github.com/pytorch/pytorch/issues/122811
